### PR TITLE
Fix docs and add residual calculation function

### PR DESCRIPTION
**Pull request by yannikschaelte**

### CODE REVIEW

File: petab/calculate.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+
+def calculate_residuals(
+    measurement_dfs: List[pd.DataFrame],
+    simulation_dfs: List[pd.DataFrame],
+    observable_dfs: List[pd.DataFrame],
+    normalize: bool = True
+) -> List[pd.DataFrame]:
+    """Calculate residuals.
+
```
Comment (dweindl) [inline petab/calculate.py]: Document arguments

File: petab/calculate.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    Returns
+    -------
+    residual_dfs:
+        Data frames in the same structure as `measurement_dfs`
+        with a field `residual` instead of measurement.
+    """
+    # convenience
+    if isinstance(measurement_dfs, pd.DataFrame):
```
Comment (dweindl) [inline petab/calculate.py]: Show people the convenience in the typehints?

File: petab/calculate.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+            # non-normalized residual is just the difference
+            residual = simulation - measurement
+
+            if normalize:
+                # look up noise standard deviation
+                observable = observable_df.loc[row[OBSERVABLE_ID]]
+                std = observable[NOISE_FORMULA]
+                try:
+                    std = float(std)
```
Comment (dweindl) [inline petab/calculate.py]: Long loop body, much nesting. Can we extract some part to a meaningful function? At least the content of the outermost loop would make a good separate function.

File: petab/calculate.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        compared_cols = set(MEASUREMENT_DF_COLS)
+        compared_cols -= {MEASUREMENT}
+        compared_cols &= set(measurement_df.columns)
+
+        # iterate over measurements, find corresponding simulations
+        for irow, row in measurement_df.iterrows():
+            measurement = row[MEASUREMENT]
+
+            # look up in simulation df
+            masks = [simulation_df[col] == row[col] for col in compared_cols]
```
Comment (dweindl) [inline petab/calculate.py]: assumes measurement_df and simulation_df have same columns? safe, I guess?

File: petab/calculate.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
+import pandas as pd
```
Comment (dweindl) [inline petab/calculate.py]: module docstring?

File: tests/test_calculate.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        OBSERVABLE_ID: ['obs_a', 'obs_a'],
+        SIMULATION_CONDITION_ID: ['c0', 'c0'],
+        TIME: [10, 10],
+        MEASUREMENT: [0, 1]
+    })
+
+    observable_df = pd.DataFrame(data={
+        OBSERVABLE_ID: ['obs_a'],
+        OBSERVABLE_FORMULA: ['A'],
+        NOISE_FORMULA: [2]
```
Comment (FFroehlich) [inline tests/test_calculate.py]: maybe add a test with non-numeric noise_formula?

File: petab/calculate.py (hunk 4 above)
Comment (yannikschaelte) [reply petab/calculate.py]: yes, should be

File: petab/calculate.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+from typing import List, Union
+
+from .C import *
+import petab
+
+
+def calculate_residuals(
+    measurement_dfs: Union[List[pd.DataFrame], pd.DataFrame],
+    simulation_dfs: Union[List[pd.DataFrame], pd.DataFrame],
+    observable_dfs: Union[List[pd.DataFrame], pd.DataFrame],
```
Comment (dweindl) [inline petab/calculate.py]: Codacy Issue found: [Wrong hanging indentation before block (add 4 spaces).](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=5076455)

File: petab/calculate.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+import pandas as pd
+from functools import reduce
+from typing import List, Union
+
+from .C import *
+import petab
+
+
+def calculate_residuals(
+    measurement_dfs: Union[List[pd.DataFrame], pd.DataFrame],
```
Comment (dweindl) [inline petab/calculate.py]: Codacy Issue found: [Wrong hanging indentation before block (add 4 spaces).](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=5076455)

File: petab/calculate.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+            measurement_dfs, simulation_dfs, observable_dfs):
+        residual_df = calculate_residuals_for_table(
+            measurement_df, simulation_df, observable_df, normalize, scale)
+        residual_dfs.append(residual_df)
+    return residual_dfs
+
+
+def calculate_residuals_for_table(
+    measurement_df: pd.DataFrame,
+    simulation_df: pd.DataFrame,
```
Comment (dweindl) [inline petab/calculate.py]: Codacy Issue found: [Wrong hanging indentation before block (add 4 spaces).](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=5076455)

File: tests/test_calculate.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
+from petab import calculate_residuals
```
Comment (dweindl) [inline tests/test_calculate.py]: Codacy Issue found: [Missing module docstring](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=5076455)

File: petab/calculate.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+    observable_df: pd.DataFrame,
```
Comment (dweindl) [inline petab/calculate.py]: Codacy Issue found: [Wrong hanging indentation before block (add 4 spaces).](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=5076455)

File: petab/calculate.py [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+    normalize: bool = True,
+    scale: bool = True
```
Comment (dweindl) [inline petab/calculate.py]: Codacy Issue found: [Wrong hanging indentation before block (add 4 spaces).](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=5076455)

File: petab/calculate.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+    for (measurement_df, simulation_df, observable_df) in zip(
+            measurement_dfs, simulation_dfs, observable_dfs):
+        residual_df = calculate_residuals_for_table(
+            measurement_df, simulation_df, observable_df, normalize, scale)
+        residual_dfs.append(residual_df)
+    return residual_dfs
+
+
+def calculate_residuals_for_table(
+    measurement_df: pd.DataFrame,
```
Comment (dweindl) [inline petab/calculate.py]: Codacy Issue found: [Wrong hanging indentation before block (add 4 spaces).](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=5076455)

File: petab/calculate.py [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+    simulation_df: pd.DataFrame,
+    observable_df: pd.DataFrame,
+    normalize: bool = True,
```
Comment (dweindl) [inline petab/calculate.py]: Codacy Issue found: [Wrong hanging indentation before block (add 4 spaces).](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=5076455)

File: petab/calculate.py [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+from functools import reduce
+from typing import List, Union
+
+from .C import *
+import petab
+
+
+def calculate_residuals(
+    measurement_dfs: Union[List[pd.DataFrame], pd.DataFrame],
+    simulation_dfs: Union[List[pd.DataFrame], pd.DataFrame],
```
Comment (dweindl) [inline petab/calculate.py]: Codacy Issue found: [Wrong hanging indentation before block (add 4 spaces).](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=5076455)

File: tests/test_calculate.py (hunk 6 above)
Comment (yannikschaelte) [reply tests/test_calculate.py]: those are currently not yet supported, just numeric replacements

File: tests/test_calculate.py (hunk 10 above)
Comment (dweindl) [inline tests/test_calculate.py]: Codacy Issue found: [Missing module docstring](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=5076455)

File: tests/test_calculate.py (hunk 6 above)
Comment (yannikschaelte) [reply tests/test_calculate.py]: addressed in https://github.com/PEtab-dev/PEtab/pull/347

**Pull request merged**